File: nightsweeper/report.py
"""Morning report (U14) — the honest artifact.

Always prints each paid lane's utilization every night, so honesty is structural,
not gated on a threshold (R20/R27). Recommends downgrading a paid lane when, over
the last ``window_nights``, its spend is below ``spend_pct_threshold`` of its
budget AND its lane-attributable passes are below ``min_passes`` — pass-per-dollar
is first-class, so a high-spend/low-pass lane also triggers (R18/AE6).
"""

from __future__ import annotations

from datetime import datetime, timedelta


def _paid_lanes(config) -> dict:
    """name → nightly_budget for lanes that cost money (unit usd)."""
    out = {}
    for b in config.backends:
        budget = b.options.get("nightly_budget")
        if budget:
            out[b.name] = float(budget)
    return out


def _window_start(night_start: str, nights: int) -> str:
    try:
        d = datetime.fromisoformat(night_start)
    except ValueError:
        return night_start
    return (d - timedelta(days=nights)).isoformat()


def downgrade_candidates(config, ledger, night_start: str) -> list:
    """Return [(lane, evidence)] for lanes that should be downgraded."""
    dg = config.report.downgrade
    win_start = _window_start(night_start, dg.window_nights)
    hist = ledger.lane_summary(win_start)
    out = []
    for lane, budget in _paid_lanes(config).items():
        s = hist.get(lane, {"consumed": 0.0, "passes": 0, "attempts": 0})
        if s["attempts"] == 0:
            continue  # never nag an unused/young lane — no evidence to act on
        budget_window = budget * dg.window_nights
        util = (s["consumed"] / budget_window) if budget_window > 0 else 0.0
        low_pass = s["passes"] < dg.min_passes
        underused = util < dg.spend_pct_threshold
        # recommend on too-few-passes when the lane is EITHER under-used (low spend)
        # OR paying-for-failure (spent money for too few passes) — pass-per-dollar first-class.
        if low_pass and (underused or s["consumed"] > 0):
            out.append((lane, {
                "window_nights": dg.window_nights,
                "spend": s["consumed"], "budget_window": budget_window,
                "utilization_pct": round(util * 100, 1), "passes": s["passes"],
            }))
    return out


def generate(config, ledger, summary, inventory: dict, night_start: str) -> str:
    paid = _paid_lanes(config)
    tonight = ledger.lane_summary(night_start)
    rows = ledger.runs_since(night_start)
    lines = [f"# Nightsweeper morning report — {night_start[:10]}", ""]

    if summary.tasks_total == 0:
        lines += ["**No backlog, no run.** No tasks were returned by any configured source.",
                  "Nightsweeper never invents work."]
        text = "\n".join(lines) + "\n"
        _write(config, text)
        return text

    lines += [
        "## Summary",
        f"- Tasks seen: **{summary.tasks_total}**",
        f"- Dispatched: **{summary.dispatched}** · Passed: **{summary.passed}** · "
        f"Parked: **{summary.parked}**",
        f"- Backlog remaining (unprocessed this night): **{summary.backlog_remaining}**",
        f"- Stop reason: `{summary.stop_reason}`",
        "",
        "## Per-lane consumption (tonight)",
        "| Lane | Attempts | Passes | $ consumed | Utilization |",
        "|---|---:|---:|---:|---|",
    ]
    for b in sorted(config.backends, key=lambda x: x.cost_rank):
        s = tonight.get(b.name, {"consumed": 0.0, "passes": 0, "attempts": 0})
        if b.name in paid:
            budget = paid[b.name]
            util = f"{round(100*s['consumed']/budget,1)}% of ${budget:.2f} budget"
            ppd = (s["passes"] / s["consumed"]) if s["consumed"] > 0 else 0.0
            util += f" · {ppd:.2f} passes/$"
            spend = f"${s['consumed']:.2f}"
        else:
            util, spend = "free ($0)", "$0.00"
        lines.append(f"| {b.name} | {s['attempts']} | {s['passes']} | {spend} | {util} |")

    lines += ["", "## Backlog"]
    parked_rows = [r for r in rows
                   if r["park_reason"] or r["validation_result"] in ("parked", "skipped")]
    lines.append(f"- Parked for human review: **{summary.parked}**")
    lines.append(f"- Bare (un-enrolled) TODO/FIXME markers (not dispatched): "
                 f"**{inventory.get('bare_todo_count', 0)}**")
    if parked_rows:
        lines.append("")
        lines.append("Parked tasks:")
        for r in parked_rows:
            reason = r["park_reason"] or r["validation_result"]
            lines.append(f"  - `{r['task_id']}` ({r['backend']}) — {reason}")

    cands = downgrade_candidates(config, ledger, night_start)
    lines += ["", "## Plan utilization & recommendations"]
    if not paid:
        lines.append("- No paid lanes configured.")
    for lane, budget in paid.items():
        s = tonight.get(lane, {"consumed": 0.0})
        lines.append(f"- `{lane}`: spent ${s['consumed']:.2f} of ${budget:.2f} tonight "
                     f"(utilization always reported, recommendation or not).")
    for lane, ev in cands:
        lines.append(
            f"- **Recommend downgrading `{lane}`.** Over the last {ev['window_nights']} nights it "
            f"used ${ev['spend']:.2f} of ${ev['budget_window']:.2f} ({ev['utilization_pct']}%) "
            f"and cleared {ev['passes']} task(s). Honesty over engagement — you are paying for "
            f"capacity you are not using."
        )

    # Dormant V2 line: only rendered once predicted_lo/hi are populated.
    if any(r["predicted_lo"] is not None for r in rows):
        lines += ["", "## Preflight accuracy (V2)",
                  "- (predicted-vs-actual bracketing line renders here when preflight is active.)"]

    text = "\n".join(lines) + "\n"
    _write(config, text)
    return text


def _write(config, text: str) -> None:
    try:
        with open(config.report.path, "w", encoding="utf-8") as f:
            f.write(text)
    except OSError:
        pass

File: lib/report.mjs
// The morning report (Node port). Honest by construction: always shows per-agent
// consumption + utilization for every paid agent, recommends a downgrade when an
// agent is paid-for-but-underused, flags adjudication-gate rejections; never inflates.
import { writeFileSync } from 'node:fs';

// name -> nightly budget (usd) for agents that actually cost money.
function paidAgents(config) {
  const out = {};
  for (const a of config.agents) {
    const b = a.options?.nightly_budget;
    if (b) out[a.name] = Number(b);
  }
  return out;
}

function windowStart(nightStart, nights) {
  const d = new Date(nightStart);
  if (Number.isNaN(d.getTime())) return nightStart;
  d.setUTCDate(d.getUTCDate() - nights);
  return d.toISOString();
}

// [(agent, evidence)] for paid agents that should be downgraded: too few passes AND
// (under-used OR paying-for-failure). Pass-per-dollar is first-class (R18). An unused
// young agent is never nagged — no attempts means no evidence.
export function downgradeCandidates(config, ledger, nightStart) {
  const dg = config.report.downgrade;
  const hist = ledger.laneSummary(windowStart(nightStart, dg.window_nights));
  const out = [];
  for (const [agent, budget] of Object.entries(paidAgents(config))) {
    const s = hist[agent] ?? { consumed: 0, passes: 0, attempts: 0 };
    if (s.attempts === 0) continue;
    const budgetWindow = budget * dg.window_nights;
    const util = budgetWindow > 0 ? s.consumed / budgetWindow : 0;
    const lowPass = s.passes < dg.min_passes;
    const underused = util < dg.spend_pct_threshold;
    if (lowPass && (underused || s.consumed > 0)) {
      out.push([agent, {
        window_nights: dg.window_nights, spend: s.consumed,
        budget_window: budgetWindow, utilization_pct: Math.round(util * 1000) / 10, passes: s.passes,
      }]);
    }
  }
  return out;
}

export function generateReport(config, ledger, summary, nightStart) {
  const paid = paidAgents(config);
  const L = [`# Nightsweeper morning report — ${nightStart.slice(0, 10)}`, ''];
  if (summary.tasks_total === 0) {
    L.push('**No backlog, no run.** No chores were ready. Nightsweeper never invents work.');
    const t = L.join('\n') + '\n';
    try { writeFileSync(config.report.path, t); } catch {}
    return t;
  }
  const tonight = ledger.laneSummary(nightStart);
  const rows = ledger.since(nightStart);
  L.push('## Summary',
    `- Chores seen: **${summary.tasks_total}**`,
    `- Done: **${summary.passed}** · Set aside (parked): **${summary.parked}** · Dispatched: **${summary.dispatched}**`,
    `- Stop reason: \`${summary.stop_reason}\``, '',
    '## Per-agent work tonight', '| Agent | Attempts | Done | $ spent | Utilization |', '|---|---:|---:|---:|---|');
  for (const a of [...config.agents].sort((x, y) => x.cost_rank - y.cost_rank)) {
    const s = tonight[a.name] ?? { attempts: 0, passes: 0, consumed: 0 };
    let util, spend;
    if (paid[a.name]) {
      const budget = paid[a.name];
      const ppd = s.consumed > 0 ? (s.passes / s.consumed).toFixed(2) : '0.00';
      util = `${Math.round((1000 * s.consumed) / budget) / 10}% of $${budget.toFixed(2)} · ${ppd} passes/$`;
      spend = `$${s.consumed.toFixed(2)}`;
    } else {
      util = 'free ($0)'; spend = '$0.00';
    }
    L.push(`| ${a.name} | ${s.attempts} | ${s.passes} | ${spend} | ${util} |`);
  }

  // Plan utilization & recommendations — always reported, recommendation or not (R20/R27).
  L.push('', '## Plan utilization & recommendations');
  if (!Object.keys(paid).length) L.push('- No paid agents configured — every agent is free.');
  for (const [agent, budget] of Object.entries(paid)) {
    const s = tonight[agent] ?? { consumed: 0 };
    L.push(`- \`${agent}\`: spent $${s.consumed.toFixed(2)} of $${budget.toFixed(2)} tonight (utilization always reported).`);
  }
  for (const [agent, ev] of downgradeCandidates(config, ledger, nightStart)) {
    L.push(`- **Recommend downgrading \`${agent}\`.** Over the last ${ev.window_nights} nights it used `
      + `$${ev.spend.toFixed(2)} of $${ev.budget_window.toFixed(2)} (${ev.utilization_pct}%) and cleared `
      + `${ev.passes} chore(s). Honesty over engagement — you are paying for capacity you are not using.`);
  }

  const gateFails = {};
  for (const r of rows) if (String(r.validation_result).startsWith('failed:gate:')) {
    const g = r.validation_result.slice('failed:gate:'.length);
    gateFails[g] = (gateFails[g] ?? 0) + 1;
  }
  if (Object.keys(gateFails).length) {
    L.push('', '## Adjudication gates');
    for (const [g, c] of Object.entries(gateFails))
      L.push(`- **${g}** rejected ${c} change(s) that passed their tests — kept out of a passing branch.`);
  }
  const parked = rows.filter((r) => r.park_reason);
  if (parked.length) {
    L.push('', '## Set aside for you');
    for (const r of parked) L.push(`  - \`${r.task_id}\` (${r.backend}) — ${r.park_reason}`);
  }

  // Preflight accuracy (V2): only renders once an agent's cost_model populates predictions.
  const pred = rows.filter((r) => r.predicted_lo != null && ['passed', 'failed'].includes(r.validation_result));
  if (pred.length) {
    const bracketed = pred.filter((r) => r.predicted_lo <= (r.consumed ?? 0) && (r.consumed ?? 0) <= r.predicted_hi).length;
    const pct = Math.round((1000 * bracketed) / pred.length) / 10;
    L.push('', '## Preflight accuracy (V2)',
      `- [predicted_lo, predicted_hi] bracketed actual cost in **${bracketed}/${pred.length} (${pct}%)** `
      + 'of predicted dispatches (≥70% supports promoting preflight from advisory to gate).');
  }
  const t = L.join('\n') + '\n';
  try { writeFileSync(config.report.path, t); } catch {}
  return t;
}
